### DISCUSSION TITLE
TRPCClientError: Unexpected end of JSON input

### DISCUSSION THREAD

**Discussion by yekta** [trpc/trpc#5777] [General]
**Update: I updated to `11.0.0-rc.396` and couldn't reproduce the problem ever since. I was on `11.0.0-rc.366` before.**

I'm calling `mutateAsync` inside a function using TRPCReact. However, sometimes (and only sometimes) I'm getting `TRPCClientError: Unexpected end of JSON input`. I confirmed that the object I'm passing to the function is a valid and that this call isn't even hitting the start of my procedure on the server (although it is hitting the server). I'm wondering if there is a way to debug this because something internal seems to be going wrong and all I can see is `Unexpected end of JSON input`. After a couple of hours of trial and error, I still couldn't pinpoint exactly what happens since my debug logs aren't resulting in anything meaningful because of where the error occurs.

```js
// trpc.ts
const trpc = createTRPCReact<AppRouter>();

// component.tsx
//..............
const mutation = trpc.runGenerator.useMutation()
const res = await mutation.mutateAsync({
    slug,
    params
})
//..............
```

Here is what I mean by not hitting my procedure:

```js
// server.ts
export const runGenerator = procedure
	.input(
		z.object({
			slug: z.string(),
			params: z.array(z.string())
		})
	)
	.mutation(async ({ input, ctx }) => {
		const start = Date.now();
		console.log(`----- 🟡 Running generator -----`);
//...........
```

That console log isn't executed which makes me think something goes wrong in the input phase but my input seems perfectly fine since it goes through when I make the request a couple more times with the exact same inputs.

Here is the response:
<img width="459" alt="Screenshot 2024-06-07 at 9 31 20 PM" src="https://github.com/trpc/trpc/assets/40706601/b0bbcd8f-8b09-4930-8263-cec262103a45">

Here is the payload:
<img width="389" alt="Screenshot 2024-06-07 at 9 33 24 PM" src="https://github.com/trpc/trpc/assets/40706601/7a5307e7-cd95-40b1-90a4-a734f81ce967">

**Comment by bnussman** [comment]
I'm also seeing this intermittent 400 errors happening on one of my mutations 😔
I'm using Bun with the tRPC's `fetchRequestHandler`.

Here is what **Sentry** shows:
<img width="1029" height="220" alt="Screenshot 2026-05-29 at 7 39 02 PM" src="https://github.com/user-attachments/assets/faa0525c-9704-4673-ab3e-99e69fe9f5dc" />

Here is the code that Sentry seems to point me to... Not sure exactly what I'm looking at here...
<img width="945" height="644" alt="Screenshot 2026-05-29 at 7 38 41 PM" src="https://github.com/user-attachments/assets/6146dfc5-8e82-4a80-aad9-7f23532143ee" />

**Reply by bnussman** [reply]
Welp... Here's my fix: https://github.com/bnussman/beep/commit/1b0a9da46909e355d4f81b0d7f5872843eabfd38

I wish tRPC had some guardrails around what I was doing wrong at the TypeScript level but maybe there is some kind of limitation I'm not aware of preventing that.
